src/fu-tool.c

```diff
@@ -1356,39 +1392,30 @@ fu_util_activate (FuUtilPrivate *priv, gchar **values, GError **error)
 		return FALSE;
 	}
 
-	/* nothing to do */
-	for (guint i = 0; i < devices->len; i++) {
-		FuDevice *dev = g_ptr_array_index (devices, i);
-		if (fu_device_has_flag (dev, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION)) {
-			fu_engine_add_plugin_filter (priv->engine,
-						     fu_device_get_plugin (dev));
-			has_pending = TRUE;
-		}
-	}
-
-	if (!has_pending) {
-    		g_printerr ("No firmware to activate\n");
-    		return TRUE;
-	}
-
-	/* load engine */
-	if (!fu_util_start_engine (priv, FU_ENGINE_LOAD_FLAG_READONLY_FS, error))
-		return FALSE;
-
 	/* activate anything with _NEEDS_ACTIVATION */
+	/* order by device priority */
+	g_ptr_array_sort (devices, fu_util_activate_task_sort_cb);
 	for (guint i = 0; i < devices->len; i++) {
 		FuDevice *device = g_ptr_array_index (devices, i);
 		if (!fu_util_filter_device (priv, FWUPD_DEVICE (device)))
 			continue;
 		if (!fu_device_has_flag (device, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION))
 			continue;
+		has_pending = TRUE;
 		/* TRANSLATORS: shown when shutting down to switch to the new version */
 		g_print ("%s %s…\n", _("Activating firmware update"), fu_device_get_name (device));
 		if (!fu_engine_activate (priv->engine, fu_device_get_id (device), error))
 			return FALSE;
 	}
 
-	return TRUE;
+	if (!has_pending) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOTHING_TO_DO,
+				     "No devices to activate");
+	}
+
+	return has_pending;
```

Comment:
I think `return TRUE` is much easier on your mental model here.

Reply:
fixed